core/src/main/java/org/elasticsearch/index/query/functionscore/FunctionScoreQueryParser.java

```diff
@@ -120,160 +102,121 @@ public Query parse(QueryShardContext context) throws IOException, QueryParsingEx
             } else {
                 while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
                     if (token == XContentParser.Token.FIELD_NAME) {
                         currentFieldName = parser.currentName();
-                    } else if (parseContext.parseFieldMatcher().match(currentFieldName, WEIGHT_FIELD)) {
+                    } else if (parseContext.parseFieldMatcher().match(currentFieldName, FunctionScoreQueryParser.WEIGHT_FIELD)) {
                         functionWeight = parser.floatValue();
                     } else {
                         if ("filter".equals(currentFieldName)) {
-                            filter = parseContext.parseInnerFilter();
+                            filter = parseContext.parseInnerFilterToQueryBuilder();
                         } else {
-                            // do not need to check null here,
-                            // functionParserMapper throws exception if parser
-                            // non-existent
-                            ScoreFunctionParser functionParser = functionParserMapper.get(parseContext, currentFieldName);
-                            scoreFunction = functionParser.parse(context, parser);
+                            scoreFunction = copyFunctionSource(parser, currentFieldName);
                         }
                     }
                 }
+
                 if (functionWeight != null) {
-                    scoreFunction = new WeightFactorFunction(functionWeight, scoreFunction);
+                    if (scoreFunction == null) {
+                        //if only weight is present, that is the score function
+                        scoreFunction = new WeightBuilder().setWeight(functionWeight).buildAsBytes(XContentType.JSON);
+                    } else {
+                        //if both weight and function are present, we need to combine them
+                        final BytesReference function = scoreFunction;
+                        scoreFunction = new ScoreFunctionBuilder() {
+                            @Override
+                            protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+                                XContentParser parser = XContentFactory.xContent(function).createParser(function);
+                                parser.nextToken();
+                                //skip the start_object token from the function, we have to copy only starting from the name of the function
+                                parser.nextToken();
+                                builder.copyCurrentStructure(parser);
+                            }
+
+                            @Override
+                            public String getName() {
+                                throw new UnsupportedOperationException();
+                            }
+                        }.setWeight(functionWeight).buildAsBytes(XContentType.JSON);
+                    }
```

Comment:
Could you explain why this logic is now more complex then before (as before it was just setting a variable)?